### PR TITLE
Add coverage in bitmap, epoch, and ringbuffer.

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Signed-off-by: Alan Jowett <alan.jowett@microsoft.com>

## Description

Add additional test coverage based on codecov data.

## Testing

New test.

## Documentation

No.

### CODE REVIEW

File: libs/platform/unit/platform_unit_test.cpp [hunk 1]
```diff
@@ -285,6 +286,74 @@ TEST_CASE("epoch_test_two_threads", "[platform]")
...
+    }
+
+  private:
+    std::mutex lock;
+    std::condition_variable condition_variable;
+    bool signaled = false;
+};
+
+TEST_CASE("epoch_test_stale_items", "[platform]")
+{
```
Comment (shankarseal) [inline libs/platform/unit/platform_unit_test.cpp:315]: Some comments pleases what this test wants to accomplish.

**Comment by Alan-Jowett** [comment]
Moving back to draft to crashes.

File: libs/platform/ebpf_epoch.h [hunk 2]
```diff
@@ -93,6 +93,16 @@ extern "C"
     void
     ebpf_epoch_free_work_item(_Frees_ptr_opt_ ebpf_epoch_work_item_t* work_item);
 
+    /**
+     * @brief Check the state of the free list on a CPU.
+     *
+     * @param[in] cpu_id CPU to check.
+     * @return true - Free list is empty.
+     * @return false - Free list is not empty.
```
Comment (dthaler) [inline libs/platform/ebpf_epoch.h:101]:
```suggestion
     * @retval true Free list is empty.
     * @retval false Free list is not empty.
```

File: libs/platform/user/ebpf_platform_user.cpp [hunk 3]
```diff
@@ -20,15 +22,145 @@ bool _ebpf_platform_is_preemptible = true;
...
+ * Work items can be queued to this thread, which then executes them with out being interrupte by lower
+ * priority threads.
+ */
+class _ebpf_emulated_dpc
+{
+  public:
+    _ebpf_emulated_dpc() = delete;
+
+    /**
+     * @brief Construct a new ebpf emulated dpc object for CPU i
```
Comment (dthaler) [inline libs/platform/user/ebpf_platform_user.cpp:52]: line is cut off in mid-word

File: libs/platform/user/ebpf_platform_user.cpp [hunk 4]
```diff
@@ -20,15 +22,145 @@ bool _ebpf_platform_is_preemptible = true;
...
+        // Insert a marker in the queue.
+        ebpf_list_initialize(&flush_entry);
+        ebpf_list_insert_tail(&head, &flush_entry);
+        condition_variable.notify_all();
+        // Wait until the marker is processed.
+        condition_variable.wait(l, [this]() { return terminate || ebpf_list_is_empty(&flush_entry); });
+    }
+
+    /**
+     * @brief Insert a work item into it's associated queue.
```
Comment (dthaler) [inline libs/platform/user/ebpf_platform_user.cpp:114]:
```suggestion
     * @brief Insert a work item into its associated queue.
```

File: libs/platform/user/ebpf_platform_user.cpp [hunk 5]
```diff
@@ -20,15 +22,145 @@ bool _ebpf_platform_is_preemptible = true;
...
+     *
+     * @param[in] work_item Work item to be enqueued.
+     * @param[in] parameter_1 Parameter to pass to worker function.
+     * @return true - Work item wasn't already queued.
+     * @return false - Work item is already queued.
```
Comment (dthaler) [inline libs/platform/user/ebpf_platform_user.cpp:119]:
```suggestion
     * @retval true Work item wasn't already queued.
     * @retval false Work item is already queued.
```

File: libs/platform/user/ebpf_platform_user.cpp [hunk 6]
```diff
@@ -20,15 +22,145 @@ bool _ebpf_platform_is_preemptible = true;
...
+     */
+    static bool
+    insert(ebpf_non_preemptible_work_item_t* work_item, void* parameter_1)
```
Comment (dthaler) [inline libs/platform/user/ebpf_platform_user.cpp:122]:
```suggestion
    insert(_In_ ebpf_non_preemptible_work_item_t* work_item, _In_opt_ void* parameter_1)
```

File: tests/unit/test.vcxproj [hunk 7]
```diff
@@ -195,6 +195,7 @@
     <ClInclude Include="..\..\libs\thunk\mock\mock.h" />
     <ClInclude Include="..\end_to_end\helpers.h" />
     <ClInclude Include="..\end_to_end\test_helper.hpp" />
+    <ClCompile Include="wer_report_test_wrapper.cpp" />
```
Comment (dthaler) [inline tests/unit/test.vcxproj:198]: also need to add it to tests/unit/CMakeLists.txt

File: tests/unit/wer_report_test_wrapper.cpp [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+
+#include "catch_wrapper.hpp"
+
+uint32_t _vectored_exception_handler_test = 1;
+
+ULONG AddVectoredExceptionHandler_test_first = 0;
+PVECTORED_EXCEPTION_HANDLER AddVectoredExceptionHandler_test_handler = nullptr;
+
+_Ret_maybenull_ PVOID
+AddVectoredExceptionHandler_test(_In_ ULONG first, _In_ PVECTORED_EXCEPTION_HANDLER handler)
```
Comment (dthaler) [inline tests/unit/wer_report_test_wrapper.cpp:34]:
```suggestion
AddVectoredExceptionHandler_test(ULONG first, _In_ PVECTORED_EXCEPTION_HANDLER handler)
```
Per SAL documentation, `_In_` is only for pointers, not value types.

File: tests/unit/wer_report_test_wrapper.cpp [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+uint32_t WerReportCreate_test_report_handle = 2;
+
+bool WerReportCreate_test_fail = false;
+bool WerReportCreate_test_called = false;
+
+HRESULT
+WINAPI
+WerReportCreate_test(
+    _In_ const wchar_t* event_type,
+    _In_ WER_REPORT_TYPE report_type,
```
Comment (dthaler) [inline tests/unit/wer_report_test_wrapper.cpp:67]:
```suggestion
    WER_REPORT_TYPE report_type,
```
Or does this have to conform to a (bad) prototype in a windows header file?

File: libs/platform/user/ebpf_platform_user.cpp [hunk 10]
```diff
@@ -20,15 +22,145 @@ bool _ebpf_platform_is_preemptible = true;
...
+ * Work items can be queued to this thread, which then executes them without being interrupted by lower
+ * priority threads.
+ */
+class _ebpf_emulated_dpc
+{
+  public:
+    _ebpf_emulated_dpc() = delete;
+
+    /**
+     * @brief Construct a new ebpf emulated dpc object for CPU i
```
Comment (dthaler) [inline libs/platform/user/ebpf_platform_user.cpp:52]:
```suggestion
     * @brief Construct a new ebpf emulated dpc object for CPU i.
```

File: libs/platform/user/ebpf_platform_user.cpp [hunk 11]
```diff
@@ -20,15 +22,145 @@ bool _ebpf_platform_is_preemptible = true;
...
+                        l.lock();
+                    }
+                }
+                condition_variable.wait(l, [this]() { return terminate || !ebpf_list_is_empty(&head); });
+            }
+        });
+    }
+
+    /**
+     * @brief Destroy the ebpf emulated dpc object
```
Comment (dthaler) [inline libs/platform/user/ebpf_platform_user.cpp:87]:
```suggestion
     * @brief Destroy the ebpf emulated dpc object.
```

File: tests/unit/wer_report_test_wrapper.cpp (hunk 9 above)
Comment (dthaler) [reply tests/unit/wer_report_test_wrapper.cpp:69]: (unresolving since I saw no change or answer to the question yet)

File: tests/unit/wer_report_test_wrapper.cpp (hunk 8 above)
Comment (Alan-Jowett) [reply tests/unit/wer_report_test_wrapper.cpp:35]: SAL needs to match the wrong annotation in the Windows header.

File: tests/unit/wer_report_test_wrapper.cpp (hunk 9 above)
Comment (Alan-Jowett) [reply tests/unit/wer_report_test_wrapper.cpp:69]: My bad. Should have mentioned that this has to match the incorrect SAL annotation in the Windows header.

**Review by dthaler — APPROVED** [review]

File: libs/platform/user/ebpf_platform_user.cpp [hunk 12]
```diff
@@ -82,16 +84,145 @@ _clean_up_thread_pool()
...
+
+thread_local static bool _ebpf_non_preemptible = false;
+
+typedef struct _ebpf_non_preemptible_work_item
+{
+    ebpf_list_entry_t entry;
+    void* context;
+    _ebpf_emulated_dpc* queue;
+    void* parameter_1;
+    void (*work_item_routine)(void* work_item_context, void* parameter_1);
```
Comment (dthaler) [inline libs/platform/user/ebpf_platform_user.cpp:97]: No SAL annotation on the function arguments?

**Review by dthaler — APPROVED** [review]

**Review by dthaler — APPROVED** [review]

File: libs/platform/unit/platform_unit_test.cpp [hunk 13]
```diff
@@ -684,9 +759,14 @@ bitmap_test()
     ebpf_bitmap_t* bitmap = reinterpret_cast<ebpf_bitmap_t*>(data.data());
     ebpf_bitmap_initialize(bitmap, bit_count);
 
-    // Set every even bit interlocked.
-    for (size_t i = 0; i < bit_count; i += 2) {
-        ebpf_bitmap_set_bit(bitmap, i, true);
+    // Set every bit.
```
Comment (shankarseal) [inline libs/platform/unit/platform_unit_test.cpp:762]: why this change?

File: libs/platform/user/ebpf_platform_user.cpp [hunk 14]
```diff
@@ -684,6 +820,9 @@ ebpf_free_timer_work_item(_Frees_ptr_opt_ ebpf_timer_work_item_t* work_item)
 
     WaitForThreadpoolTimerCallbacks(work_item->threadpool_timer, true);
     CloseThreadpoolTimer(work_item->threadpool_timer);
+    for (auto& dpc : _ebpf_emulated_dpcs) {
```
Comment (shankarseal) [inline libs/platform/user/ebpf_platform_user.cpp:823]: I dont understand why is this needed here?

File: libs/platform/unit/platform_unit_test.cpp (hunk 13 above)
Comment (Alan-Jowett) [reply libs/platform/unit/platform_unit_test.cpp:762]: This is to cover the set and reset interlocked and non-interlocked bitmap operations.

File: libs/platform/user/ebpf_platform_user.cpp (hunk 14 above)
Comment (Alan-Jowett) [reply libs/platform/user/ebpf_platform_user.cpp:823]: Epoch code relies on a KeFlushDpc call when freeing a timer. This is to provide similar functionality.

File: libs/platform/user/ebpf_platform_user.cpp [hunk 15]
```diff
@@ -82,16 +84,145 @@ _clean_up_thread_pool()
...
+    /**
+     * @brief Insert a work item into its associated queue.
+     *
+     * @param[in] work_item Work item to be enqueued.
+     * @param[in] parameter_1 Parameter to pass to worker function.
+     * @retval true Work item wasn't already queued.
+     * @retval false Work item is already queued.
+     */
+    static bool
+    insert(_In_ ebpf_non_preemptible_work_item_t* work_item, _In_opt_ void* parameter_1)
```
Comment (shankarseal) [inline libs/platform/user/ebpf_platform_user.cpp:184]: consider making this an instance method.

**Review by dthaler — APPROVED** [review]

**Review by dthaler — APPROVED** [review]

**Review by shankarseal — APPROVED** [review]
